docs/functional/architecture/0009-Compiled-Backend-Integration.md

```diff
@@ -0,0 +1,79 @@
+# Integration of compiled backends
+
+- **Status**: valid 
+- **Authors**: Peter Kardos (@petiaccja)
+- **Created**: 2022-07-13
+- **Updated**: 2022-07-13
+
+Summary of the key design choices made for the live execution of generated C++ (and other compiled) code.
+
+## Context
+
+As per the current state, gt4py can execute iterator IR by generating Python code from it which is loaded back as a module. Additionally, gt4py can also emit equivalent C++ code from iterator IR. The integration of the compiled backends focuses on compiling the emitted C++ code and calling it from Python, effectively executing iterator IR live by translating it to machine code.
+
+The process of executing ITIR this way consists of the following steps:
+1. Generate ITIR
+2. Generate C++ code equivalent (*fencil code*) to ITIR
+3. Generate C++ code that wraps the fencil code and can be accessed from Python (*binding code*)
+4. Compile the fencil and binding codes into machine code (a dynamically linked library)
+5. Load the dynamic library into Python and extract the interfaces as Python objects 
+6. Call the interfaces from Python with the fencil's arguments
```

Comment:
Yeah, maybe we should have a separate section for this whole types and arguments thing, because the current implementation is just not correct.

Version a): we treat the execution as dynamically typed, and you can only call it with concrete arguments and we recompile when needed
Version b): we treat the execution as statically typed, and you must compile it with a set of types and then call it with only those types

What we have now basically is that you can only call it with concrete arguments, and the first time you do it determines the types, and then you can only call it with the same types, which of course works for field view but is sketchy in general.

Reply:
Changed caching so now it's fully dynamically typed.